### PR TITLE
[checkpoint][dt] Add SavePlanner for Distributed Tensor

### PR DESCRIPTION
**Pull request by wz337**
[T128869838]

This PR include the following:
1. Implements DT checkpoint through a DistributedTensorSavePlanner and DistributedTensorLoadPlanner. 
2. To test the planner, we create a module with a sharded-dt param and a replicated-dt param and use it to test the saving and loading behavior. 
3. Add checkpoint_utils to add a wrapper to initialize a temp directory for checkpoint testing.

### CODE REVIEW

**Review by kumpera — CHANGES_REQUESTED** [review]

File: spmd/dt_save_planner.py [hunk 1]
```diff
@@ -0,0 +1,285 @@
...
+class DistributedTensorSavePlanner(DefaultSavePlanner):
+    def create_local_plan(self) -> SavePlan:
+        requests = []
+        for fqn, obj in self.state_dict.items():
+            if isinstance(obj, DTensor):
+                tensor = obj
+                offsets, sizes, dedup_replicated_tensors = get_box_for(
+                    tensor, tensor.device_mesh.get_coordinate_on_dim(0)
+                )
+                self.dedup_replicated_tensors = dedup_replicated_tensors
```
Comment (kumpera) [inline spmd/dt_save_planner.py:100]: What happens if the second tensor is sharded?

File: spmd/dt_save_planner.py [hunk 2]
```diff
@@ -0,0 +1,285 @@
...
+
+    offsets = [val * idx for val in offsets]
+    return (torch.Size(offsets), size, tensor.placements[0].is_replicate())
+
+
+class DistributedTensorSavePlanner(DefaultSavePlanner):
+    def create_local_plan(self) -> SavePlan:
+        requests = []
+        for fqn, obj in self.state_dict.items():
+            if isinstance(obj, DTensor):
```
Comment (kumpera) [inline spmd/dt_save_planner.py:95]: This needs an else clause to handle non-dt stuff.

Comment (wz337) [inline spmd/dt_save_planner.py:95]: This doesn't handle non-DT write items.
Add else.

File: spmd/checkpoint/dt_planner.py [hunk 3]
```diff
@@ -0,0 +1,207 @@
...
+    logger.propagate = False
+    return logger
+
+
+dLogger: logging.Logger = init_logger()
+
+
+class DistributedTensorSavePlanner(DefaultSavePlanner):
+    def __init__(self, dedup_replicated_tensors: bool = True):  # pyre-ignore[3]
+        self.dedup_replicated_tensors: bool = dedup_replicated_tensors
```
Comment (kumpera) [inline spmd/checkpoint/dt_planner.py:57]: should call ``super(DistributedTensorSavePlanner).__init__()``

File: test/spmd/checkpoint/test_dt_planner.py [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+)
+from spmd.testing.common_utils import (
+    TEST_GPU_NUM,
+    DistTensorTestBase,
+    with_comms,
+)
+from torch.testing._internal.common_distributed import skip_if_lt_x_gpu
+from torch.testing._internal.common_utils import run_tests
+
+CHECKPOINT_DIR = "/scratch/checkpoint/test_dt_planner/"
```
Comment (kumpera) [inline test/spmd/checkpoint/test_dt_planner.py:25]: For a test suite we should use a temporary generated name. For example: https://github.com/pytorch/pytorch/pull/84952

**Review by kumpera — APPROVED** [review]
LGTM.
Fixing the test directory should be done as it will lead to unreliable tests otherwise.

**Review by kumpera — APPROVED** [review]

**Pull request closed without merging**